### PR TITLE
[R-package] allow access to params in Booster

### PR DESCRIPTION
**Pull request by jameslamb**
I've been trying to check LightGBM questions on Stack Overflow more frequently. A search today led me to this question from two years ago: https://stackoverflow.com/questions/49674112/accessing-lightgbm-model-parameters/50115384#50115384.

It pointed out that there's no way to access the parameters from `Booster` object. This is something that the Python package has had for a while.

Storing the parameters in the Booster can be helpful for cases where you've saved a model and then later load it in a session that doesn't have the original training code available. For example, you might want to be sure that the parameters you use for training continuation are the same as those you used in the original training. Or you might want to use parameters as metadata in a model monitoring system.

This pull request adds a new public attribute, `Booster$params`. Just like `Booster.params` in the Python package, it includes all passed-in parameters AND any parameters on the Dataset used for training (such as `max_bin`).

This PR also adds the first set of tests on `saveRDS.lgb.booster()` and `readRDS.lgb.booster()`, which were previously not covered by any tests.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
(Shouldn't prevent to merge this PR) but I just want to say that a more general fix will be loading params from cpp side: https://github.com/microsoft/LightGBM/issues/2613#issuecomment-562216072

File: src/c_api.cpp [hunk 1]
```diff
@@ -294,13 +294,15 @@ class Booster {
...
+    if (param.count("num_class") && new_config.num_class != config_.num_class) {
       Log::Fatal("Cannot change num_class during training");
     }
-    if (param.count("boosting")) {
+    if (param.count("boosting") && new_config.boosting != config_.boosting) {
       Log::Fatal("Cannot change boosting during training");
     }
-    if (param.count("metric")) {
+    if (param.count("metric") && new_config.metric != config_.metric) {
       Log::Fatal("Cannot change metric during training");
```
Comment (jameslamb) [inline src/c_api.cpp:306]: I originally had this PR only changing the R package, but then ran into this error

> Error: [LightGBM] [Fatal] Cannot change metric during training

This is thrown even if you aren't actually CHANGING `metric`. I think the change here in `c_api` is closer to the desired behavior, only throwing an error if the parameter is being changed.

File: R-package/tests/testthat/test_custom_objective.R [hunk 2]
```diff
@@ -39,6 +39,8 @@ param <- list(
 )
 num_round <- 10L
 
+# this is failing because of the checks in src/c_api.cpp#294,
+# which says "if you pass params into reset-params, throw an error"
```
Comment (jameslamb) [inline R-package/tests/testthat/test_custom_objective.R:43]:
```suggestion
```

**Review by guolinke — APPROVED** [review]

File: src/c_api.cpp (hunk 1 above)
Comment (jameslamb) [reply src/c_api.cpp:309]: /gha run r-valgrind

Comment (jameslamb) [reply src/c_api.cpp:309]: https://github.com/microsoft/LightGBM/actions/runs/458548423

Comment (jameslamb) [reply src/c_api.cpp:309]: ✅ valgrind is happy
